File: src/main/java/edu/psu/ist/plato/kaiming/Entry.java
package edu.psu.ist.plato.kaiming;

import java.util.Comparator;

public abstract class Entry implements Comparable<Long> {
    public abstract long index();

    @Override
    public final int compareTo(Long l) {
        return (int) (index() - l);
    }
    
    public static Comparator<Entry> comparator = new Comparator<Entry> () {
        @Override
        public final int compare(Entry e1, Entry e2) {
            return Long.signum(e1.index() - e2.index());
        }
    };
    
    public static int searchIndex(final Entry[] entries, long index) {
        int left = 0, right = entries.length - 1;
        while (left <= right) {
            int pivot = (right + left) / 2;
            int ret = entries[pivot].compareTo(index);
            if (ret == 0)
                return pivot;
            else if (ret < 0) {
                left = pivot + 1;
            } else {
                right = pivot - 1;
            }
        }
        return -1;
    }
    
    public static Entry search(Iterable<Entry> entries, long index) {
        for (Entry e : entries) {
            if (e.index() == index)
                return e;
        }
        return null;
    }
}

File: src/main/scala/edu/psu/ist/plato/kaiming/Entry.scala
package edu.psu.ist.plato.kaiming

import utils.Indexed

abstract class Entry[A <: Arch] extends Indexed {
  
  override final def hashCode = index.hashCode()
  override final def equals(that: Any) = 
    that.isInstanceOf[AnyRef] && (this eq that.asInstanceOf[AnyRef])
  
}

abstract class MachEntry[A <: MachArch] extends Entry[A] {
  
  def mach: Machine[A]
  
  final def isTerminator: Boolean = this.isInstanceOf[Terminator[A]]
  final def asTerminator = this.asInstanceOf[Terminator[A]]

}

trait Terminator[A <: MachArch] {
  self : MachEntry[A] =>
  def isIndirect: Boolean
  def isReturn: Boolean
  def isCall: Boolean
  final def isInterprocedural: Boolean = isCall || isReturn
  final def isIntraprocedural = !isInterprocedural
  def isTargetConcrete: Boolean
  def isConditional: Boolean
  def targetIndex: Long
  def relocate(target: MachBBlock[A]): Unit
  def relocatedTarget: Option[MachBBlock[A]]
}